### PR TITLE
posix: pthread: implement pthread_getconcurrency()

### PR DESCRIPTION
**Pull request by cfriedt**
Zephyr must support all functionality of the `XSI_THREADS_EXT` subprofiling option group in order to claim it supports that subprofiling option group.

The XSI_THREADS_EXT option group is required for PSE51, PSE52, PSE53, and PSE54 Application Environment Profiles.

The `XSI_THREADS_EXT` option group is critical to be able to run POSIX threads with statically allocated thread stacks, which has been a feature of the implementation since it was initially added.
    
The `pthread_getconcurrency()` and `pthread_setconcurrency()` functions are the only remaining, unimplemented functions of the XSI_THREADS_EXT option group, as per IEEE 1003.1-2017.
    
Implement `pthread_getconcurrency()` and `pthread_setconcurrency()` via the more "posixly correct" interpretation of the specification.
    
I.e. as the pthread_t:k_thread relationship is 1:1 and not M:N, Zephyr does not support multiplexing of user threads on top of schedulable kernel entities (i.e. "user threads" are directly mapped to native threads, just like linuxthreads or NPTL are in Linux).

For that reason, to be "posixly correct", we should save the provided value via `pthread_setconcurrency()`, in the absence of errors, and also return that same value back via `pthread_getconcurrency()`, even though that serves zero purpose in Zephyr for the foreseeable future.
    
Note: the specification also states
> an implementation can always ignore any calls to pthread_setconcurrency() and return a constant for pthread_getconcurrency().

For that reason, the implementation may be revisited at a later time when when considering optimizations and when there is a better system in place for documenting deviations.

Any such optimization should be explicitly controlled via Kconfig.
    
See https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_getconcurrency.html for additional info

### CODE REVIEW

**Comment by cfriedt** [comment]
Not somehow seems to tag this as being related to CMSIS? Weird.

**Comment by kartben** [comment]
> Bot somehow tags this as being related to CMSIS? Weird.

Ya this is wrong:

https://github.com/zephyrproject-rtos/zephyr/blob/dec10e8d3dfcbcf971f6325c90f26d7d554b44a0/MAINTAINERS.yml#L476

**Review by ycsin — DISMISSED** [review]

**Comment by cfriedt** [comment]
> > Bot somehow tags this as being related to CMSIS? Weird.
> 
> Ya this is wrong:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/dec10e8d3dfcbcf971f6325c90f26d7d554b44a0/MAINTAINERS.yml#L476

Created a separate PR to fix that #65702

File: lib/posix/pthread.c
Comment (nashif) [inline lib/posix/pthread.c:464]: how is this an implementation? Doesn't it make more sense to leave this as Not supported and not implement it this way? If I have the API available and I can call it, the expectation is that it will work as specified...

File: lib/posix/pthread.c
Comment (nashif) [inline lib/posix/pthread.c:469]: It is not compliant if for example someone call this with a negative value, i.e. 

```
ERRORS
       pthread_setconcurrency() can fail with the following error:

       EINVAL new_level is negative.
```

**Review by nashif — DISMISSED** [review]

**Comment by kartben** [comment]
Just removed the trivial bug (again :)) as I don't think it's the right thing to do to try and get small-but-not-necessarily-so-trivial changes needlessly rushed

File: lib/posix/pthread.c
Comment (cfriedt) [reply lib/posix/pthread.c:464]: @nashif From https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_getconcurrency.html
> APPLICATION USAGE
> Application developers should note that an implementation can always ignore any calls to pthread_setconcurrency() and return a constant for pthread_getconcurrency(). For this reason, it is not recommended that portable applications use this function.

So, it technically does work as specified.

I'm not 100% sure if the alternative is possible today in Zephyr (i.e. adjusting at runtime the number of active threads to be less than the number of cores / harts). So it would be a long rabbit hole to go down.

File: lib/posix/pthread.c
Comment (cfriedt) [reply lib/posix/pthread.c:469]: Ok - I can fix that.

**Comment by cfriedt** [comment]
> Just removed the trivial bug (again :)) as I don't think it's the right thing to do to try and get small-but-not-necessarily-so-trivial changes needlessly rushed 

Heh - "needlessly" 😂 I'm on PTO from work this week to do other work and there are only so many hours in the day. Plus solo parenting of 3 as my wife is out of country.

But this is quite trivial in terms of conformance.

Added @andyross and @peter-mitsis, since I believe there is no way to reduce hardware concurrency at runtime in Zephyr. I could be wrong though.

**Comment by kartben** [comment]
> > Just removed the trivial bug (again :)) as I don't think it's the right thing to do to try and get small-but-not-necessarily-so-trivial changes needlessly rushed
> 
> Heh - "needlessly" 😂 I'm on PTO from work this week to do other work and there are only so many hours in the day. Plus solo parenting of 3 as my wife is out of country.

Sure, but this really has nothing to do with the PR being marked as trivial or not and the review process, no?
FWIW if the intent is to get it "fast tracked" for whatever reason (potential for merge conflict, etc.) then I think the right approach would be to ping people on #pr-help so that it has all the reviews it needs to get switfly merged as soon as the 2 required business days have passed.

**Review by ycsin — APPROVED** [review]

**Comment by cfriedt** [comment]
For clarification, I added you for review because the bot assigned 0 reviewers (?? 🤷 ) and there were 2 lines of documentation changes.

>> try and get small-but-not-necessarily-so-trivial changes needlessly rushed

A) How are these changes "not-necessarily-trivial"?

* Certainly, the 2 lines of .rst that were changed were trivial (I'm not the docs maintainer - that's just a guess).
* Certainly, the 2 lines of test code that were changed were trivial (maintainer's stamp on that)
* Certainly (I would say with > 90% certainty), the 10 lines of code that were changed were trivial (also maintainer's stamp on that)

The code changes are trivial w.r.t. the POSIX specification. The only uncertainty was about the Zephyr scheduler. I just happen to be collaborator there too, and gave an educated guess. I reviewed our scheduler API and am pretty confident about that guess now.

B) Why do you feel that merging this PR with some expedience is "needless"?

There is a need. It's blocking future work in this exact area that I maintain. I was also rushing to spend time with my 2 youngest kids. So yea... there was a need 👍

Deferring it would pretty well blow the time budget I had to work on these features for the foreseeable future.

> Sure, but this really has nothing to do with the PR being marked as trivial or not and the review process, no?

It does have something to do with it. Trivial changes should be merged faster, thus unblocking future work in this area and thus maximizing maintainer bandwidth.

> The Zephyr Project recognizes that reviewers and maintainers have limited bandwidth.

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#reviewer-expectations

I understand, if you don't have time to review this trivial change, then don't review it. But with all due respect, you shouldn't be removing the Trivial tag if the maintainer / assigneee of an area sets it. I don't set it for all of my changes - just the Trivial ones 😉. And I certainly wouldn't unset it 

> Before escalation of technical disagreements, follow the steps below:
> Resolve in the PR among assignee, maintainers and reviewer.
> Assignee to act as moderator if applicable.

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation

So ... I'm moderating here. Bear with me 😁 

> FWIW if the intent is to get it "fast tracked" for whatever reason (potential for merge conflict, etc.) then I think the right approach would be to ping people on #pr-help so that it has all the reviews it needs to get switfly merged as soon as the 2 required business days have passed.

I can understand that is what you think, but it is not explicitly mentioned anywhere in the documentation. It would be better to refer to a specific place in the process to clarify instead of waving the process flag based on a feeling. The only mention I can see that is relevant to this situation is here - maybe my Google skills are failing me, so please feel free to link to a more accurate paragraph.

https://docs.zephyrproject.org/latest/project/project_roles.html#merge-criteria

> Release engineers may make exceptions for areas with contributions primarily coming from one organisation and where reviews from other organisations are not possible, however, merges shall be completed by a person from a different organisation. In such cases, the minimum review period of at least 2 days shall be strictly followed to allow for additional reviews.

However, it is clear that "reviews from other organisations" are possible (as they have already happened). So there is is no need to be strict about waiting at least 2 days (especially for a trivial change that should have been merged within 4 hours).

With all due respect. I'm going to jump on the "limited bandwidth" wagon (because a lot of mine was spent needlessly here) and mark the changes as trivial because they are. It's not just my opinion, but the POSIX spec says it in plain language. Please don't undo that this time unless you can find some technical evidence to back it up.

@henrikbrixandersen - I know you love process as well, so please feel free to point at a more accurate paragraph if you still want to support Benjamin.

File: lib/posix/pthread.c
Comment (cfriedt) [reply lib/posix/pthread.c:464]: Also, POSIX requires that all components of a particular option group are supported if the implementation is to claim that the option group is supported. Alternatively, if they are to be unimplemented, the symbols still need to be present, but docs must state that they fall into the category of undefined behaviour.

These were the only two missing functions from `XSI_THREADS_EXT`. The trivial implementation conforms to the spec as well, so there is no need to document any kind of undefined behaviour.

Screenshot 2023-11-23 at 10 20 39 PM

If Zephyr is to one day support the functionality for adjusting concurrency dynamically, then we can revisit, but that will not be anytime soon.

**Review by ycsin — DISMISSED** [review]

**Review by carlescufi — DISMISSED** [review]

File: lib/posix/pthread.c
Comment (nashif) [reply lib/posix/pthread.c:464]: If you are going to add this implementation that does not have any effect you should not document this as supported. You need to say that this has no effect and implemented for completeness and source code compatbility. This is not the case now. See an example of how this is done elsewhere:

https://www.qnx.com/developers/docs/6.6.0_anm11_wf10/#com.qnx.doc.neutrino.lib_ref/topic/p/pthread_getconcurrency.html

> The .... RTOS doesn't support the multiplexing of user threads on top of several kernel scheduled entities. As such, the [pthread_setconcurrency()](https://www.qnx.com/developers/docs/6.6.0_anm11_wf10/com.qnx.doc.neutrino.lib_ref/topic/p/pthread_setconcurrency.html) and pthread_getconcurrency() functions are provided for source code compatibility but they have no effect when called. To maintain the function semantics, the new_level parameter is saved when pthread_setconcurrency() is called so that a subsequent call to pthread_getconcurrency() returns the same value.

File: doc/services/portability/posix/option_groups/index.rst
Comment (nashif) [inline doc/services/portability/posix/option_groups/index.rst:88]: so, IMO you need to document this differently here, maybe as posix implementation continues to grow we need a better way to document specifics, not a table.

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/index.rst:88]: Correct. I brought the same thing up in my previous PR that addressed POSIX docs. It's out of scope of the current PR though.

**Comment by nashif** [comment]
under which category does this fall exactly?

>[Trivial](https://docs.zephyrproject.org/3.5.0/project/dev_env_and_tools.html#trivial)
Trivial changes are those that appear obvious enough and do not require maintainer or code-owner involvement. Such changes should not change the logic or the design of a subsystem or component. For example a trivial change can be:
>>Documentation changes
>> Configuration changes
>>Minor Build System tweaks
>>Minor optimization to code logic without changing the logic
>>Test changes and fixes
>>Sample modifications to support additional configuration or boards etc.

Even if any of those apply, a change marked as trivial can be contested  through review where trivial becomes void and the PR need to follow the standard process. Trivial should never be used to fast track PRs, it is just a way for us to deal with low hanging fruits that are obvious enough to be merged quickly without any thorough review or any further discussion. We are past this point already :)

File: lib/posix/pthread.c
Comment (cfriedt) [reply lib/posix/pthread.c:464]: Again - reworking posix docs are out of the scope of this PR.

Comment (nashif) [reply lib/posix/pthread.c:464]: I am not asking you to rework the documentation here. 

The change to the documentation related to those calls however is wrong. This is not supported and misleading to say this function is supported when you are just adding it for source code compatibility. So please remove the 'supported' and come up with a new way to document such functions without reworking the docs or add a footnote a section somewhere indicating this and similar calls hav no effect etc and implemented for source code compatibility only.

File: doc/services/portability/posix/option_groups/index.rst
Comment (nashif) [reply doc/services/portability/posix/option_groups/index.rst:88]: sure, I said "maybe" indicating this is something to look into in the future. My point here is that this change to the docs is still wrong.

**Comment by cfriedt** [comment]
@nashif - I have revised the implementation as per your request, although as per the spec, it was not necessary.

I will not be restructuring the POSIX documentation in this PR.

**Review by ycsin — DISMISSED** [review]

**Comment by cfriedt** [comment]
> under which category does this fall exactly?
> 
> >[Trivial](https://docs.zephyrproject.org/3.5.0/project/dev_env_and_tools.html#trivial)
> Trivial changes are those that appear obvious enough and do not require maintainer or code-owner involvement.

Yes - if anyone had made the pr in its original form, I would have approved and I would have been fine if anyone else had approved it.

> Such changes should not change the logic or the design of a subsystem or component.

Exactly - it's still the same situation: set concurrency - no effect. Just now it's more complicated than it needs to be. I would have been perfectly fine with an enhancement once a similar kernel enhancement was made.

> Trivial should never be used to fast track PRs

Correct. Trivial is used to merge trivial PRs which this was originally. Now it's a more complex version of useless, but it's what you requested 🤷🏼‍♂️

> We are past this point already :)

Certainly are - will you approve finally?

**Comment by cfriedt** [comment]
* rebased to resolve merge conflict

**Review by ycsin — APPROVED** [review]
Refresh +1

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by nashif** [comment]
> @nashif - I have revised the implementation as per your request, although as per the spec, it was not necessary.
> 



I have not asked for a full implementation of something that can't be supported. I would have been fine with an empty stub and a different way of documenting this, i.e. `s/supported/implemented with no effect/` or whatever.

> I will not be restructuring the POSIX documentation in this PR.

I never asked for this, expanding the documentation however to cover above cases will be neceassry to be clear about the level of support and setting expectation right.  I would have been fine with a footnote, but adding plain **supported** in there is not correct IMO.


> > under which category does this fall exactly?
> > > [Trivial](https://docs.zephyrproject.org/3.5.0/project/dev_env_and_tools.html#trivial)
> > > Trivial changes are those that appear obvious enough and do not require maintainer or code-owner involvement.
> 
> Yes - if anyone had made the pr in its original form, I would have approved and I would have been fine if anyone else had approved it.
> 


Ok, but another reviewer coming in and requesting the Trivial to be removed is fully ok and we need to respect that and revert to normal review process.

> > Such changes should not change the logic or the design of a subsystem or component.
> 
> Exactly - it's still the same situation: set concurrency - no effect. Just now it's more complicated than it needs to be.

This is new code and a new implementation of an API, even if empty. 


image

Please read the following which we have recently adopted: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation

> If no progress is made, the assignee (maintainer) has the right to dismiss stale, unrelated or irrelevant change requests by reviewers giving the reviewers a minimum of 1 business day to respond and revisit their initial change requests or start the escalation process.

If you have any issues with the above, please raise them as a topic in the process meeting and we can discuss this.

**Comment by dleach02** [comment]
Folks, this PR broke CI.

/__w/zephyr/zephyr/lib/posix/pthread.c:477:16: error: 'ret' may be used uninitialized [-Werror=maybe-uninitialized]

**Comment by cfriedt** [comment]
> Folks, this PR broke CI.
> 
> /__w/zephyr/zephyr/lib/posix/pthread.c:477:16: error: 'ret' may be used uninitialized [-Werror=maybe-uninitialized]
> 
> 

It's a pretty simple fix.

Kind of annoying that it somehow managed to pass CI before being merged.

**Pull request merged**
